src/compiler/checker.ts

```diff
@@ -3950,6 +3987,24 @@ namespace ts {
                         return [symbol];
                     }
                 }
+
+                function sortByBestName(a: Symbol, b: Symbol) {
+                    if (some(a.declarations, hasNonGlobalAugmentationExternalModuleSymbol) && some(b.declarations, hasNonGlobalAugmentationExternalModuleSymbol)) {

```

Comment:
This does a lot of calculation on every comparison. It might be better to precompute some of the repetitive work ahead of time, something like:

```ts
const parentSpecifiers = parents!.map(symbol => 
    some(symbol.declarations, hasNonGlobalAugmentationExternalModuleSymbol)
        ? getSpecifierForModuleSymbol(symbol, context) 
        : undefined);
const indices = parents!.map((_, i) => i);
indices.sort(sortByBestName);
const sortedParts = indices.map(i => parents![i]);

...
function sortByBestName(a: number, b: number) {
  const specifierA = parentSpecifiers[a];
  const specifierB = parentSpecifiers[b];
  if (specifierA && specifierB) {
    // rest of sortByName goes here
  }
  return 0;
}
```

Reply:
Is `| undefined` necessary? `sortByBestName` doesn't check for definite assignment.